# Week 2 — Day 3: Linear Algebra for Machine Learning

This notebook introduces the main linear algebra concepts used in machine learning:

- Vectors
- Matrices
- Dot products
- Matrix multiplication
- Shape compatibility

The examples use NumPy to show how data samples and model calculations can be represented using vectors and matrices.

## 1. Why Linear Algebra Matters in Machine Learning

Machine learning models work mainly with numbers arranged as vectors and matrices.

A single data sample can be represented as a vector, while a complete dataset can be represented as a matrix. Model parameters are also stored as vectors or matrices.

Operations such as the dot product and matrix multiplication allow a model to process many features and samples efficiently.

## 2. Importing NumPy

NumPy provides the array structures and mathematical operations needed to work with vectors and matrices efficiently.

In [1]:
import numpy as np

## 3. Vectors

A vector is an ordered collection of numbers.

In machine learning, a vector often represents one data sample. Each value inside the vector represents one feature.

In the following example, one student is represented using:

1. Study hours
2. Completed projects
3. Practice tests

In [2]:
feature_names = ["study_hours", "projects_completed", "practice_tests"]
student = np.array([6, 3, 4])
print("Features:", feature_names)
print("Student vector:", student)
print("Shape:", student.shape)
print("Number of dimensions:", student.ndim)

Features: ['study_hours', 'projects_completed', 'practice_tests']
Student vector: [6 3 4]
Shape: (3,)
Number of dimensions: 1


### Interpretation

The student is represented by a one-dimensional vector containing three features.

Its shape is `(3,)`, meaning that it contains three values arranged in one dimension.

## 4. Matrices

A matrix is a two-dimensional collection of numbers arranged in rows and columns.

In a machine learning dataset:

- Each row represents one data sample.
- Each column represents one feature.

The following matrix represents three students using the same three features.

In [3]:
students = np.array([
    [6, 3, 4],
    [4, 2, 3],
    [8, 4, 5]
])
print("Students matrix:")
print(students)
print("\nShape:", students.shape)
print("Number of dimensions:", students.ndim)

Students matrix:
[[6 3 4]
 [4 2 3]
 [8 4 5]]

Shape: (3, 3)
Number of dimensions: 2


### Interpretation

The matrix has a shape of `(3, 3)`.

This means that the dataset contains three students and three features for each student.

## 5. The Dot Product

The dot product multiplies matching values from two vectors and then adds the results.

In machine learning, it is commonly used to combine feature values with model weights.

For two vectors:

`features = [6, 3, 4]`

`weights = [2.0, 4.0, 1.5]`

The dot product is calculated as:

`(6 × 2.0) + (3 × 4.0) + (4 × 1.5)`

In [4]:
student = np.array([6, 3, 4])
weights = np.array([2.0, 4.0, 1.5])
manual_result = (
    student[0] * weights[0]
    + student[1] * weights[1]
    + student[2] * weights[2]
)
numpy_result = np.dot(student, weights)
print("Manual calculation:", manual_result)
print("NumPy dot product:", numpy_result)

Manual calculation: 30.0
NumPy dot product: 30.0


### Interpretation
Both calculations produced the same result: `30.0`.
The NumPy dot product performs the same multiplication-and-sum process as the manual calculation, but in a shorter and more efficient form.

## 6. Producing Predictions for Multiple Samples

A model can calculate one result for every sample by multiplying the complete data matrix by a weight vector.

A bias value can then be added to each result.

The simplified calculation is:

`prediction = features · weights + bias`

The weights and bias in this example are only used to demonstrate the calculation. They do not represent a trained model.

In [5]:
weights = np.array([2.0, 4.0, 1.5])
bias = 5
predictions = students @ weights + bias
print("Predictions:", predictions)

Predictions: [35.  25.5 44.5]


### Interpretation

The matrix multiplication produced one result for each student:

- Student 1: `35.0`
- Student 2: `25.5`
- Student 3: `44.5`

The output has three values because the input matrix contains three rows.

### Verifying One Prediction

Matrix multiplication calculates the dot product for every row in the dataset.

To verify this, we can calculate the prediction for the first student separately using `np.dot()` and compare it with the first value produced by the matrix calculation.

In [6]:
first_student_prediction = np.dot(students[0], weights) + bias
print("Prediction calculated separately:", first_student_prediction)
print("First matrix prediction:", predictions[0])
print("Results match:", np.isclose(first_student_prediction, predictions[0]))

Prediction calculated separately: 35.0
First matrix prediction: 35.0
Results match: True


### Interpretation

The separate calculation and the first matrix prediction both produced `35.0`.

This confirms that matrix multiplication applies the same dot-product calculation to every row in the dataset, producing one prediction for each sample.

## 7. Understanding Shape Compatibility

Matrix multiplication requires compatible shapes.

The students matrix has the shape:

`(3, 3)`

The weight vector has the shape:

`(3,)`

The three columns in the matrix match the three values in the weight vector. Therefore, the multiplication is valid and produces one result for each of the three rows.

The final output shape is:

`(3,)`

The general shape rule is:

`(m, n) @ (n,) → (m,)`

Here:

- `m` is the number of samples.
- `n` is the number of features.

The inner dimensions must match because each feature needs a corresponding weight.

## 8. Shape Mismatch Example

A shape mismatch occurs when the number of matrix columns does not match the number of values in the weight vector.

The next example intentionally uses an incorrect weight vector to demonstrate the error safely.

In [7]:
wrong_weights = np.array([2.0, 4.0])

try:
    students @ wrong_weights
except ValueError as error:
    print("Shape mismatch error:")
    print(error)

Shape mismatch error:
matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 2 is different from 3)


### Why the Error Occurred

The students matrix contains three features, but the incorrect weight vector contains only two weights.

The multiplication cannot match every feature with a corresponding weight.

To fix the problem, the weight vector must contain three values.

## 9. Fixing the Shape Mismatch

The error can be fixed by creating a weight vector with one weight for each feature.

In [8]:
correct_weights = np.array([2.0, 4.0, 1.5])
fixed_result = students @ correct_weights
print("Matrix shape:", students.shape)
print("Weights shape:", correct_weights.shape)
print("Result:", fixed_result)

Matrix shape: (3, 3)
Weights shape: (3,)
Result: [30.  20.5 39.5]


## 10. Summary
In this notebook:
- A vector was used to represent one data sample.
- A matrix was used to represent multiple samples.
- The dot product was calculated manually and with NumPy.
- Matrix multiplication was used to calculate results for multiple samples at once.
- Array shapes were inspected to understand multiplication compatibility.
- A shape mismatch was demonstrated and fixed.
These operations form the basic mathematical structure behind many machine learning models, including linear regression and logistic regression.